assets/js/base/context/cart-checkout/shipping/index.js

```diff
@@ -154,36 +121,71 @@ export const ShippingDataProvider = ( { children } ) => {
 	// emit events
 	// @todo add emitters for shipping rate selection.
 	useEffect( () => {
-		if ( ! shippingOptionsLoading && currentErrorStatus.hasError ) {
+		if ( ! shippingRatesLoading && currentErrorStatus.hasError ) {
 			emitEvent(
 				currentObservers.current,
-				EMIT_TYPES.SHIPPING_RATES_SUCCESS,
+				EMIT_TYPES.SHIPPING_RATES_FAIL,
 				shippingErrorStatus
 			);
-		} else if ( ! shippingOptionsLoading && shippingOptions ) {
+		} else if ( ! shippingRatesLoading && shippingRates ) {
 			emitEvent(
 				currentObservers.current,
 				EMIT_TYPES.SHIPPING_RATES_SUCCESS,
-				shippingOptions
+				shippingRates
 			);
 		}
 	}, [
-		shippingOptions,
-		shippingOptionsLoading,
+		shippingRates,
+		shippingRatesLoading,
 		currentErrorStatus,
 		shippingErrorStatus,
 	] );
 
+	//emit shipping rate selection events
+	useEffect( () => {
+		if ( ! isSelectingRate && currentErrorStatus.hasError ) {
+			emitEvent(
+				currentObservers.current,
+				EMIT_TYPES.SHIPPING_RATE_SELECT_FAIL,
+				shippingErrorStatus
+			);
+		} else if ( ! isSelectingRate && selectedRates ) {
+			emitEvent(
+				currentObservers.current,
+				EMIT_TYPES.SHIPPING_RATE_SELECT_SUCCESS,
+				selectedRates
+			);
+		}
+	}, [
+		selectedRates,
+		isSelectingRate,
+		currentErrorStatus,
+		shippingErrorStatus,
+	] );
+
+	// dispatch checkout error if there's a shipping error.
+	useEffect( () => {
+		if ( currentErrorStatus.hasError ) {
+			dispatchActions.setHasError();
+		} else {
+			dispatchActions.clearError();
+		}
+	}, [
+		currentErrorStatus,
+		dispatchActions.setHasError,
+		dispatchActions.clearError,
+	] );
```

Comment:
This is a bit iffy.  There's _potential_ for a race condition if there are other things that might trigger checkout error (i.e. validation error/payment error etc).  We may want checkout error state to be something that is controlled in a child component (`CheckoutProcessing`?) that can set the checkout error based on a combination of any known error statuses from the other providers (shipping/billing/payment-method, validation etc).

Reply:
So move hasError to the CheckoutProvider, since it's the outermost provider? Is that what you mean?

Reply:
No, I mean in the `CheckoutProcessor`. The checkout processor component will have access to all the error states from the various providers and could use those to derive (and update) the checkout provider error state (so there'd be one place that checkout error state is maintained).

I'm still on the fence with this though, I think we could merge this and test how things work once everything else is wired up.